gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/integration/use_case/DeleteIngestedApisUseCase.java

```diff
@@ -0,0 +1,163 @@
+        PlanQueryService planQueryService,
+        SubscriptionQueryService subscriptionQueryService,
+        CloseSubscriptionDomainService closeSubscriptionDomainService,
+        DeleteSubscriptionDomainService deleteSubscriptionDomainService,
+        DeletePlanDomainService deletePlanDomainService,
+        PageQueryService pageQueryService,
+        DeleteApiDocumentationDomainService deleteApiDocumentationDomainService,
+        AuditDomainService auditDomainService,
+        ApiMetadataDomainService apiMetadataDomainService,
+        DeleteMembershipDomainService deleteMembershipDomainService,
+        ApiCrudService apiCrudService,
+        ApiIndexerDomainService apiIndexerDomainService
+    ) {
+        this.apiQueryService = apiQueryService;
+        this.planQueryService = planQueryService;
+        this.subscriptionQueryService = subscriptionQueryService;
+        this.closeSubscriptionDomainService = closeSubscriptionDomainService;
+        this.deleteSubscriptionDomainService = deleteSubscriptionDomainService;
+        this.deletePlanDomainService = deletePlanDomainService;
+        this.pageQueryService = pageQueryService;
+        this.deleteApiDocumentationDomainService = deleteApiDocumentationDomainService;
+        this.auditDomainService = auditDomainService;
+        this.apiMetadataDomainService = apiMetadataDomainService;
+        this.deleteMembershipDomainService = deleteMembershipDomainService;
+        this.apiCrudService = apiCrudService;
+        this.apiIndexerDomainService = apiIndexerDomainService;
+    }
+
+    public void execute(Input input) {
+        var apisToDelete = apiQueryService
+            .search(
+                ApiSearchCriteria.builder().integrationId(input.integrationId).build(),
+                null,
+                ApiFieldFilter.builder().pictureExcluded(true).definitionExcluded(true).build()
+            )
+            .filter(api -> !api.getApiLifecycleState().equals(Api.ApiLifecycleState.PUBLISHED))
+            .toList();
+
+        apisToDelete.forEach(api -> delete(api, input.auditInfo));
+    }
+
+    @Builder
+    public record Input(String integrationId, AuditInfo auditInfo) {}
+
+    private void delete(Api api, AuditInfo auditInfo) {
+        var apiId = api.getId();
+
+        //Get plans
+        var plans = planQueryService.findAllByApiId(apiId);
+
+        //Close active Subscriptions
+        plans.forEach(plan ->
+            subscriptionQueryService
+                .findActiveSubscriptionsByPlan(plan.getId())
+                .forEach(activeSubscription -> closeSubscriptionDomainService.closeSubscription(activeSubscription.getId(), auditInfo))
+        );
+
+        //Delete all subscriptions
+        plans.forEach(plan ->
+            subscriptionQueryService
+                .findSubscriptionsByPlan(plan.getId())
+                .forEach(subscription -> deleteSubscriptionDomainService.delete(subscription, auditInfo))
+        );
```

Comment:
What about merge the 3 foreach on plans?


```java
plans.forEach(plan -> {
            subscriptionQueryService
                .findActiveSubscriptionsByPlan(plan.getId())
                .forEach(activeSubscription -> closeSubscriptionDomainService.closeSubscription(activeSubscription.getId(), auditInfo));
        
        subscriptionQueryService
        .findSubscriptionsByPlan(plan.getId())
        .forEach(subscription -> deleteSubscriptionDomainService.delete(subscription, auditInfo));
        }

        deletePlanDomainService.delete(plan, auditInfo);
        );
```

Reply:
👍